nautobot/core/api/views.py

```diff
@@ -263,6 +263,32 @@ def dispatch(self, request, *args, **kwargs):
             self.logger.warning(msg)
             return self.finalize_response(request, Response({"detail": msg}, status=409), *args, **kwargs)
 
+    def list(self, request, *args, **kwargs):
+        """Special-case handling for CSV format as it needs different data to be passed to the Response."""
+        if "text/csv" in request.accepted_media_type:
+            queryset = self.filter_queryset(self.get_queryset())
+            serializer_class = self.get_serializer_class()
+            filename = f"{settings.BRANDING_PREPENDED_FILENAME}{queryset.model._meta.verbose_name_plural}.csv"
+            return Response(
+                data={"queryset": queryset, "serializer_class": serializer_class},
+                headers={"Content-Disposition": f"attachment; filename={filename}"},
+            )
+
+        return super().list(request, *args, format=format, **kwargs)
+
+    def retrieve(self, request, *args, **kwargs):
+        """Special-case handling for CSV format as it needs different data to be passed to the Response."""
+        if "text/csv" in request.accepted_media_type:
+            instance = self.get_object()
+            serializer_class = self.get_serializer_class()
+            filename = f"{settings.BRANDING_PREPENDED_FILENAME}{instance._meta.verbose_name}.csv"
+            return Response(
+                data={"instance": instance, "serializer_class": serializer_class},
+                headers={"Content-Disposition": f"attachment; filename={filename}"},
+            )
+
+        return super().retrieve(request, *args, format=format, **kwargs)
```

Comment:
The main reason here is actually so that I can pass the instance/queryset and the serializer_class to the renderer. By default a renderer only gets the `data` dict derived from the serializer.

It's possible that it might be doable to revise this pattern, as the biggest reason for this was to be able to convert foreign keys and the like from PK/URL strings to natural-keys - since in the latest diffs I've added `natural_key_slug` as a field on BaseModelSerializer, we **might** be able to just get everything from the serializer data at depth 1?

Reply:
Oh right, it's also to bypass the pagination that the default `list()` implementation will enforce.

Reply:
Looks like maybe I can/should override our `paginate_queryset()` implementation instead, will do some further digging.